File: backend/app/candle/subscription_manager.py
from starlette.websockets import WebSocket

SubKey = tuple[str, int]


class CandleSubscriptionManager:
    def __init__(self) -> None:
        self._clients_by_key: dict[SubKey, set[WebSocket]] = {}

    def subscribe(self, symbol: str, interval_minutes: int, client: WebSocket) -> bool:
        """Returns True if this is the (symbol, interval)'s first subscriber (0 -> 1)."""
        key = (symbol, interval_minutes)
        clients = self._clients_by_key.setdefault(key, set())
        is_new = len(clients) == 0
        clients.add(client)
        return is_new

    def unsubscribe(self, symbol: str, interval_minutes: int, client: WebSocket) -> bool:
        """Returns True if this was the (symbol, interval)'s last subscriber (-> 0)."""
        key = (symbol, interval_minutes)
        clients = self._clients_by_key.get(key)
        if not clients or client not in clients:
            return False
        clients.discard(client)
        if not clients:
            del self._clients_by_key[key]
            return True
        return False

    def unsubscribe_all(self, client: WebSocket) -> list[SubKey]:
        """Called on client disconnect. Returns keys that lost their last subscriber."""
        emptied = []
        for key in list(self._clients_by_key.keys()):
            symbol, interval_minutes = key
            if self.unsubscribe(symbol, interval_minutes, client):
                emptied.append(key)
        return emptied

    def clients_for(self, symbol: str, interval_minutes: int) -> set[WebSocket]:
        return self._clients_by_key.get((symbol, interval_minutes), set())

File: backend/internal/candle/subscription_manager.go
package candle

import "sync"

type SubKey struct {
	Symbol          string
	IntervalMinutes int
}

type SubscriptionManager struct {
	mu           sync.Mutex
	clientsByKey map[SubKey]map[*ClientState]struct{}
}

func NewSubscriptionManager() *SubscriptionManager {
	return &SubscriptionManager{clientsByKey: make(map[SubKey]map[*ClientState]struct{})}
}

// returns true if this is first subscriber
func (m *SubscriptionManager) Subscribe(symbol string, intervalMinutes int, client *ClientState) bool {
	m.mu.Lock()
	defer m.mu.Unlock()
	key := SubKey{symbol, intervalMinutes}
	clients, ok := m.clientsByKey[key]
	if !ok {
		clients = make(map[*ClientState]struct{})
		m.clientsByKey[key] = clients
	}
	isNew := len(clients) == 0
	clients[client] = struct{}{}
	return isNew
}

// returns true if this was the last subscriber
func (m *SubscriptionManager) Unsubscribe(symbol string, intervalMinutes int, client *ClientState) bool {
	m.mu.Lock()
	defer m.mu.Unlock()
	key := SubKey{symbol, intervalMinutes}
	clients, ok := m.clientsByKey[key]
	if !ok {
		return false
	}
	if _, present := clients[client]; !present {
		return false
	}
	delete(clients, client)
	if len(clients) == 0 {
		delete(m.clientsByKey, key)
		return true
	}
	return false
}

// returns keys that got unsubscribed
func (m *SubscriptionManager) UnsubscribeAll(client *ClientState) []SubKey {
	m.mu.Lock()
	keys := make([]SubKey, 0, len(m.clientsByKey))
	for key := range m.clientsByKey {
		keys = append(keys, key)
	}
	m.mu.Unlock()

	var emptied []SubKey
	for _, key := range keys {
		if m.Unsubscribe(key.Symbol, key.IntervalMinutes, client) {
			emptied = append(emptied, key)
		}
	}
	return emptied
}

func (m *SubscriptionManager) ClientsFor(symbol string, intervalMinutes int) []*ClientState {
	m.mu.Lock()
	defer m.mu.Unlock()
	clients := m.clientsByKey[SubKey{symbol, intervalMinutes}]
	result := make([]*ClientState, 0, len(clients))
	for c := range clients {
		result = append(result, c)
	}
	return result
}
